File: occasum-core/src/main/java/pw/stamina/occasum/scan/field/StandardFieldLocatorService.java
package pw.stamina.occasum.scan.field;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public final class StandardFieldLocatorService implements FieldLocatorService {

    @Override
    public List<Field> findAllFields(Object source, FieldIgnorePredicate ignorePredicate) {
        List<Field> fields = findAllFields(source.getClass());

        fields.removeIf(ignorePredicate);
        fields.forEach(this::ensureAccessibility);

        return fields;
    }

    private List<Field> findAllFields(Class<?> type) {
        List<Field> fields = new ArrayList<>();

        for (; type != null; type = type.getSuperclass()) {
            fields.addAll(Arrays.asList(type.getDeclaredFields()));
        }

        return fields;
    }

    private void ensureAccessibility(Field field) {
        if (!field.isAccessible()) {
            field.setAccessible(true);
        }
    }
}

File: occasum-core/src/main/kotlin/pw/stamina/occasum/scan/field/StandardFieldLocatorService.kt
package pw.stamina.occasum.scan.field

import java.lang.reflect.Field
import java.util.*

class StandardFieldLocatorService : FieldLocatorService {

    override fun findAllFields(source: Any, ignorePredicate: FieldIgnorePredicate): Collection<Field> {
        val fields = findAllFields(source.javaClass)

        val filteredFields = fields.filter(ignorePredicate)

        filteredFields.forEach(this::ensureAccessibility)

        return fields
    }

    private fun findAllFields(type: Class<*>): Collection<Field> {
        val fields = ArrayList<Field>()
        var type: Class<*>? = type

        while (type != null) {
            fields.addAll(type.declaredFields.toList())
            type = type.superclass
        }

        return fields
    }

    private fun ensureAccessibility(field: Field) {
        if (!field.isAccessible) {
            field.isAccessible = true
        }
    }
}
